src/new/spawnmanager/spawnmanager_hooks.cpp

```diff
@@ -0,0 +1,127 @@
+/*******************************************************************************
+/*                 O P E N  S O U R C E  --  V I N I F E R A                  **
+/*******************************************************************************
+ *
+ *  @project       Vinifera
+ *
+ *  @file          SPAWNMANAGER_HOOKS.CPP
+ *
+ *  @author        CCHyper
+ *
+ *  @brief         Contains the hooks for SpawnManagerClass
+ *                 and KamikazeTrackerClass.
+ *
+ *  @license       Vinifera is free software: you can redistribute it and/or
+ *                 modify it under the terms of the GNU General Public License
+ *                 as published by the Free Software Foundation, either version
+ *                 3 of the License, or (at your option) any later version.
+ *
+ *                 Vinifera is distributed in the hope that it will be
+ *                 useful, but WITHOUT ANY WARRANTY; without even the implied
+ *                 warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
+ *                 PURPOSE. See the GNU General Public License for more details.
+ *
+ *                 You should have received a copy of the GNU General Public
+ *                 License along with this program.
+ *                 If not, see <http://www.gnu.org/licenses/>.
+ *
+ ******************************************************************************/
+
+#include "spawnmanager_hooks.h"
+
+#include "extension.h"
+#include "hooker.h"
+#include "hooker_macros.h"
+#include "spawnmanager.h"
+#include "techno.h"
+#include "technoext.h"
+#include "kamikazetracker.h"
+#include "veinholemonster.h"
+#include "vinifera_globals.h"
+
+
+/**
+ *  Patch to make the spawn manager abandon its target when ordered to idle.
+ *
+ *  @author: ZivDero
+ */
+DECLARE_PATCH(_EventClass_Execute_IDLE_Spawn_Manager_Patch)
+{
+    GET_REGISTER_STATIC(TechnoClass*, techno, esi);
+    static TechnoClassExtension* extension;
+
+    extension = Extension::Fetch<TechnoClassExtension>(techno);
+    if (extension->SpawnManager)
+        extension->SpawnManager->Abandon_Target();
+
+    static RTTIType rtti = techno->Kind_Of();
+    if (rtti == RTTI_UNIT)
+    {
+        JMP(0x00494AC5);
+    }
+    else
+    {
+        JMP(0x00495110);
+    }
+
+}
+
+
+/**
+ *  Patch to block vehicles from moving if they're currently preparing to spawn.
+ *
+ *  @author: ZivDero
+ */
+DECLARE_PATCH(_DriveLocomotionClass_Start_Of_Move_Spawn_Manager_Patch)
+{
+    GET_REGISTER_STATIC(TechnoClass*, linked_to, eax);
+    static TechnoClassExtension* extension;
+
+    _asm pushad
+
+    if (linked_to->EMPFramesRemaining > 0)
+    {
+        _asm popad
+        // return 1;
+        JMP(0x0047FE39);
+    }
+
+    extension = Extension::Fetch<TechnoClassExtension>(linked_to);
+    if (extension->SpawnManager && extension->SpawnManager->Preparing_Count())
+    {
+        _asm popad
+        // return 1;
+        JMP(0x0047FE39);
+    }
```

Comment:
I can't imagine the performance improvement from this being significant enough to warrant the code duplication, especially as it'd be in the relatively uncommon "EMP'd unit" case. And I wouldn't be surprised if the compiler ended up generating the same code anyway - avoiding fetching the extension unless necessary.

```suggestion
    extension = Extension::Fetch<TechnoClassExtension>(linked_to);
    if (linked_to->EMPFramesRemaining > 0 || (extension->SpawnManager && extension->SpawnManager->Preparing_Count()))
    {
        _asm popad
        // return 1;
        JMP(0x0047FE39);
    }
```